src/config/include/config/Node.hpp

```diff
@@ -0,0 +1,444 @@
+        typename O,
+        typename... Args,
+        typename = std::enable_if_t<
+            !isNodeConstructible<O, Args...>() &&
+            std::is_same_v<decltype(YAML::convert<O>::encode(std::declval<O>())), YAML::Node>>,
+        typename = void,
+        typename = void>
+    std::optional<O> _maybeImpl(Args&&... args) const {
+        static_assert(sizeof...(args) == 0,
+                      "Cannot pass additional args when using built-in YAML conversion");
+        return std::make_optional<O>(_yaml.as<O>());
+    }
+
+    static YAML::Node parse(std::string);
+
+    template <typename K>
+    std::optional<const YAML::Node> yamlGet(const K& key) const {
+        if (!_valid) {
+            return std::nullopt;
+        }
+        const YAML::Node found = _yaml[key];
+        if (found) {
+            return std::make_optional<const YAML::Node>(found);
+        } else {
+            if (!_parent) {
+                return std::nullopt;
+            }
+            return _parent->yamlGet(key);
+        }
+    }
+
+    template <typename K>
+    const Node get(const K& key) const {
+        const std::string keyStr = v1::toString(key);
+        if (!_valid) {
+            return Node{YAML::Node{}, this, false, keyStr};
+        }
+        if constexpr (std::is_convertible_v<K, std::string>) {
+            if (key == "..") {
+                // this is...not the most succinct business-logic ever....
+                std::stringstream childKey;
+                if (_parent) {
+                    childKey << _parent->_key << "/";
+                }
+                childKey << _key << "/..";
+                if (!_parent) {
+                    return Node{YAML::Node{}, nullptr, false, childKey.str()};
+                }
+                return Node{_parent->_yaml, _parent->_parent, _parent->_valid, childKey.str()};
+            }
+        }
+        auto yaml = this->yamlGet(key);
+        if (yaml) {
+            return Node{*yaml, this, true, keyStr};
+        } else {
+            return Node{YAML::Node{}, this, false, keyStr};
+        }
+    }
+
+    void appendKey(std::ostringstream& out) const;
+
+    YAML::Node _yaml;
+    std::string _key;
+    const Node* _parent;
+    bool _valid;
+};
+
+
+// we can act like a Node if iterated value is a scalar or
+// we can act like a pair of Nodes if iterated value is a map entry
+class IteratorValue : public std::pair<Node, Node>, public Node {
+    using NodePair = std::pair<Node, Node>;
+
+public:
+    // jump through immense hoops to avoid knowing anything about the actual yaml iterator other
+    // than its pair form is a pair of {YAML::Node, YAML::Node}
+    template <typename ITVal>
+    IteratorValue(const Node* parent, ITVal itVal, size_t index)
+        : NodePair{std::make_pair(
+              Node{itVal.first,
+                   parent,
+                   itVal.first,
+                   itVal.first ? (itVal.first.template as<std::string>() + "$key") : ""},
```

Comment:
Could `_key` be made optional or be set to an empty string here? It seems like the first item in the pair really wants to be a scalar and doesn't want to key.

Reply:
We can't really make the key optional because we don't know if we're iterating over a sequence or a map.
This does set the key to the empty-string if we're iterating over a sequence rather than a map. This gets around the `YAML::Node` oddity that the iterator for all nodes is both a node and a kvp of nodes but the k in the kvp form for arrays is basically undefined and it's not really obvious what the `kvp.first.path()` should be when iterating over maps (hence the `$key` suffix).

But I actually just changed this to 

1. add some docs because it's not super clear what this logic is intending to do
2. make the keys for these weird situations a bit more logical
3. test the behavior of `genny::Node` in these situations versus the behavior of `YAML::Node`.

Hopefully the changed version is a bit clearer!

Reply:
Gotcha! `_key` is for traversing back up to the parent. 

I should've read the unit tests.